src/cpp/ext/filters/otel/otel_client_filter.cc

```diff
@@ -0,0 +1,212 @@
+}
+
+grpc_core::ArenaPromise<grpc_core::ServerMetadataHandle>
+OpenTelemetryClientFilter::MakeCallPromise(
+    grpc_core::CallArgs call_args,
+    grpc_core::NextPromiseFactory next_promise_factory) {
+  auto* path = call_args.client_initial_metadata->get_pointer(
+      grpc_core::HttpPathMetadata());
+  auto* call_context = grpc_core::GetContext<grpc_call_context_element>();
+  auto* tracer = grpc_core::GetContext<grpc_core::Arena>()
+                     ->ManagedNew<OpenTelemetryCallTracer>(
+                         path != nullptr ? path->Ref() : grpc_core::Slice(),
+                         grpc_core::GetContext<grpc_core::Arena>());
+  GPR_DEBUG_ASSERT(
+      call_context[GRPC_CONTEXT_CALL_TRACER_ANNOTATION_INTERFACE].value ==
+      nullptr);
+  call_context[GRPC_CONTEXT_CALL_TRACER_ANNOTATION_INTERFACE].value = tracer;
+  call_context[GRPC_CONTEXT_CALL_TRACER_ANNOTATION_INTERFACE].destroy = nullptr;
+  return next_promise_factory(std::move(call_args));
+}
+
+//
+// OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer
+//
+
+OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::
+    OpenTelemetryCallAttemptTracer(OpenTelemetryCallTracer* parent,
+                                   bool arena_allocated)
+    : parent_(parent),
+      arena_allocated_(arena_allocated),
+      start_time_(absl::Now()) {
+  // TODO(yashykt): Figure out how to get this to work with absl::string_view
+  OTelPluginState().client.attempt.started->Add(
+      1, {{std::string(OTelMethodKey()), std::string(parent_->method_)}});
+}
+
+void OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::RecordSendMessage(
+    const grpc_core::SliceBuffer& send_message) {
+  RecordAnnotation(
+      absl::StrFormat("Send message: %ld bytes", send_message.Length()));
+}
+
+void OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::
+    RecordSendCompressedMessage(
+        const grpc_core::SliceBuffer& send_compressed_message) {
+  RecordAnnotation(absl::StrFormat("Send compressed message: %ld bytes",
+                                   send_compressed_message.Length()));
+}
+
+void OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::
+    RecordReceivedMessage(const grpc_core::SliceBuffer& recv_message) {
+  RecordAnnotation(
+      absl::StrFormat("Received message: %ld bytes", recv_message.Length()));
+}
+
+void OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::
+    RecordReceivedDecompressedMessage(
+        const grpc_core::SliceBuffer& recv_decompressed_message) {
+  RecordAnnotation(absl::StrFormat("Received decompressed message: %ld bytes",
+                                   recv_decompressed_message.Length()));
+}
+
+void OpenTelemetryCallTracer::OpenTelemetryCallAttemptTracer::
+    RecordReceivedTrailingMetadata(
+        absl::Status status, grpc_metadata_batch* /*recv_trailing_metadata*/,
+        const grpc_transport_stream_stats* transport_stream_stats) {
+  absl::InlinedVector<std::pair<std::string, std::string>, 2> attributes = {
```

Comment:
I assume that it's still a TODO to support additional labels here?

Reply:
yeah, when we start supporting GCP/GSM Observability, we would add that.